src/cardsort/analysis.py

```diff
@@ -13,19 +14,47 @@
 ]
 
 
-def _get_distance_matrix_for_user(df_user: pd.DataFrame) -> np.ndarray:
-    n = len(df_user)
+logger = logging.getLogger(__name__)
 
-    X = np.zeros((n, n))
-    for i in range(n):
-        for j in range(i, n):
-            cat1 = df_user.loc[df_user["card_id"] == (i + 1), "category_label"].values[
-                0
-            ]
-            cat2 = df_user.loc[df_user["card_id"] == (j + 1), "category_label"].values[
-                0
-            ]
-            X[i, j] = X[j, i] = 0 if cat1 == cat2 else 1
+
+def _check_data(df: pd.DataFrame) -> bool:
+    # check if user_ids go from 1 to n=number of users
+    user_ids = df["user_id"].unique()
+    for i, user_id in enumerate(user_ids):
+        if user_id != i + 1:
+            logger.error("User_ids are not in the right order or contain gaps.")
+            return False
+    # check if each card_id is always associated with exactly one card_label
```

Comment:
A more compact version for this code can be:
```python
    card_id_counts = df.groupby("card_id")["card_label"].nunique()
    if (card_id_counts != 1).any():
        logger.error("At least one card_id associated with different card_labels.")
```
but your version has the advantage to give a card_id associated with more than one label, whereas mine only says "There's an issue", so you can keep yours :).

Reply:
Ok, I will keep it as is for now